### PR TITLE
CLDC-1911 Add form definition for 23/24 lettings

### PR DESCRIPTION
**Pull request by kosiakkatrina**
This PR contains programmatically defined sections, subsections, pages and questions for lettings logs form that were generated from 22/23 lettings log JSON.

## Generating using rake task
The files were generated using rake task that is now removed, because there was also some manual changes to adjust generated files
- Files after running the rake task and linting: [Initially generated and linted form](https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/pull/1254/commits/f346f29f9971d334ef496b4d01f0899282240061)
-  Rake task that was run [Remove generate form task and service](https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/pull/1254/commits/555d2771cc2015c5a1acf630dff458e7837aa879)

## Manual adjustments
After running the rake task some questions were missing, because they were defined twice in JSON (for different pages) and the IDs of the questions were used for the names of the files, so those files got overriden.
To fix that I've added some of the questions manually, and some of them by running the rake task for a specific question id only and adjusting the filenames.

The list of duplicated questions that were added and their count:
"rsnvac"=>2,  different options for first let ✅                                                                                                           
 "offered"=>2,  - different hint text and header for social let  ✅                                                                                                           
 "voiddate"=>2, - different hint and header based on new build ✅
 "tenancy"=>2, - different header and hint text depending on starter tenancy ✅
 "tenancyother"=>2, - same content, no need to change ✅
 "pregnancy_value_check"=>36, - same content no need to change ✅
 "ethnic"=>5, - different options based on ethnic group ✅
 "retirement_value_check"=>16, - 2 different versions ✅
 "reason"=>2, - different header and hint text and answer options + conditional for other depending on renewal ✅
 "prevten"=>2, - different hint text and answer options depending on renewal ✅
 "referral"=>4, - different options depending on Prp and supported housing ✅
 "is_carehome"=>4, - same ✅content, no need to change
 "chcharge"=>4, - different suffix and header based on period ✅
 "brent"=>4, - depending on period ✅
 "scharge"=>4,  - depending on period ✅
 "pscharge"=>4,- depending on period ✅
 "supcharg"=>4,- depending on period ✅
 "tcharge"=>4,- depending on period ✅
 "rent_value_check"=>2 - same content, no need to change ✅

The updates to these are on separate commits for each field

Might be worth testing routing around these questions a bit more closely.

## Similar questions
Repeating questions and pages for people 2-8 were made generic by passing in person_index. Lead tenant questions are defined separately, as the content on them is different.

This also includes some soft validations that are run per person.

Each set of pages/questions is a separate commit

## Tests
The tests cover sections and subsections, so that we can keep the consistent list of subsections within sections and pages within subsections

Tests also cover person questions/pages that were changed from a set of similar questions/pages to generic questions/pages

## Running the form locally
I chose not to create a 2023/2024 form as part of this PR, because I don't think we have a mechanism to test/add future forms(there's a validation on logs being no longer than 14 days in the future) and this PR is large enough.

This means that a form using the new form elements in the PR is never initialised.

To locally initialise the form from these files for 2022/23 instead of 23/24 so that it could be tested, we can undo this change and start the server as usual:
[Undo for and form handler changes](https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/pull/1254/commits/5ea602a71687cbd688c56f1c3027f8ef1e46862c)

### CODE REVIEW

File: app/services/spike/form_generator_service.rb [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        def questions
+          @questions ||= #{questions_array}
+        end
+      end")
+      out_file.close
+    end
+
+    def create_question(question)
+      out_file = File.new("app/models/form/lettings/questions/#{question.id}.rb", "w")
+      out_file.puts("class Form::Sales::Questions::#{question.id.camelize} < ::Form::Question
```
Comment (arfacamble) [inline app/services/spike/form_generator_service.rb]: minor point - Lettings not Sales here

File: app/services/spike/form_generator_service.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    def create_section(section)
+      subsections_array = section.subsections.map { |s| "Form::Lettings::Subsections::#{section.id.camelize}::#{s.id.camelize}.new(nil, nil, self)" }
+      out_file = File.new("app/models/form/lettings/sections/#{section.id}.rb", "w")
+      out_file.puts("class Form::Lettings::Sections::#{section.id.camelize} < ::Form::Section
+  def initialize(id, hsh, form)
+    super
+    @id = \"#{section.id}\"")
+        out_file.puts("    @label = \"#{section.id}\"") if section.id
+        out_file.puts("    @description = \"#{section.description}\"") if section.description
+        out_file.puts("    @form = form
```
Comment (arfacamble) [inline app/services/spike/form_generator_service.rb]: I think these lines are indented one too many ! lulz

File: app/models/form/lettings/pages/access_needs_exist.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Form::Lettings::Pages::AccessNeedsExist < ::Form::Page
+  def initialize(id, hsh, subsection)
+    super
+    @id = "access_needs_exist"
+    @header = ""
+    @description = ""
```
Comment (kosiakkatrina) [inline app/models/form/lettings/pages/access_needs_exist.rb]: Will probably want to remove these

File: app/models/form/lettings/pages/access_needs_exist.rb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+class Form::Lettings::Pages::AccessNeedsExist < ::Form::Page
+  def questions
+    @questions ||= [Form::Lettings::Questions::Housingneeds.new(nil, nil, self)]
```
Comment (ghost) [inline app/models/form/lettings/pages/access_needs_exist.rb:3]: Housingneeds -> HousingNeeds?

File: app/models/form/lettings/pages/armed_forces.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class Form::Lettings::Pages::ArmedForces < ::Form::Page
+  def questions
+    @questions ||= [Form::Lettings::Questions::Armedforces.new(nil, nil, self)]
```
Comment (ghost) [inline app/models/form/lettings/pages/armed_forces.rb:3]: ArmedForces

File: app/models/form/lettings/pages/care_home_4_weekly.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class Form::Lettings::Pages::CareHome4Weekly < ::Form::Page
+  def initialize(id, hsh, subsection)
+    super
+    @id = "care_home_4_weekly"
+    @depends_on = [{ "period" => 3, "needstype" => 2, "household_charge" => 0 }, { "period" => 3, "needstype" => 2, "household_charge" => nil }]
+  end
+
+  def questions
+    @questions ||= [Form::Lettings::Questions::IsCarehome.new(nil, nil, self), Form::Lettings::Questions::Chcharge4Weekly.new(nil, nil, self)]
```
Comment (ghost) [inline app/models/form/lettings/pages/care_home_4_weekly.rb]: IsCareHome? i think it's missing one word when upcasing

File: app/models/form/lettings/pages/armed_forces.rb (hunk 5 above)
Comment (kosiakkatrina) [reply app/models/form/lettings/pages/armed_forces.rb:3]: Hmm, good point. It generated the class name from the question ID where it's just one word, so it didn't upcase it the same as page. Happy to change it!

File: app/models/form/lettings/pages/care_home_4_weekly.rb (hunk 6 above)
Comment (kosiakkatrina) [reply app/models/form/lettings/pages/care_home_4_weekly.rb]: Same as above, but happy to change if it's clearer!

**Review by jamesrose — APPROVED** [review]

**Pull request merged**